File: leetcode/python/p1408.py
from typing import List


class Solution:
    """
    https://leetcode.com/problems/string-matching-in-an-array
    """

    def stringMatching(self, words: List[str]) -> List[str]:
        words.sort(key=len)
        res = []
        for i in range(len(words)):
            for j in range(i + 1, len(words)):
                if words[i] in words[j]:
                    res.append(words[i])
                    break
        return res


if __name__ == "__main__":
    s = Solution()
    print(s.stringMatching(["mass", "as", "hero", "superhero"]))

File: leetcode/1000/p1408.cpp

class Solution {
   public:
    std::vector<std::string> stringMatching(std::vector<std::string>& words) {
        std::sort(words.begin(), words.end(), [](const std::string& a, const std::string& b) {
            return a.length() < b.length();
        });

        std::vector<std::string> res;
        for (int i = 0; i < words.size(); i++) {
            for (int j = i + 1; j < words.size(); j++) {
                if (words[j].find(words[i]) != std::string::npos) {
                    res.push_back(words[i]);
                    break;
                }
            }
        }
        return res;
    }
};
